driver-core/src/main/com/mongodb/client/model/expressions/ArrayExpression.java

```diff
@@ -51,28 +50,53 @@
      */
     <R extends Expression> ArrayExpression<R> map(Function<? super T, ? extends R> in);
 
-    /**
-     * Performs a reduction on the elements of this array, using the provided
-     * identity value and an associative reducing function, and returns
-     * the reduced value. The initial value must be the identity value for the
-     * reducing function.
-     *
-     * @param initialValue the identity for the reducing function
-     * @param in the associative reducing function
-     * @return the reduced value
-     */
-    T reduce(T initialValue, BinaryOperator<T> in);
```

Comment:
Can you explain how the below methods cover all the use cases of reduce?  It's not at all clear to me.

Reply:
The returned value of the binary operator `in` needs to come from some combination of its parameters `(a, b)`, and the only way to "combine" them is to use some binary method expressible in our language. I went through the list of [all methods we are presently implementing](https://docs.google.com/spreadsheets/d/1mqkG2XJ4Telnhmi1V19TpwWCjuy4jelNAfuIm9HqPcA/edit?pli=1#gid=1430738632) and considered whether each one was a reductive operation (column D). If it was, I implemented a method for it below.

Therefore every usage that involves reduction either can be modelled using one of the below, or if it cannot be modelled with one of the below, it cannot be modelled using this removed `reduce` operation regardless.

Reply:
I have a counterexample:

```
> db.test1.find()
[
  {
    a: [ { x: 3, y: 4 }, { x: 5, y: 11 } ]
  },
  {
    a: []
  }
]

> db.test1.aggregate([{$project : {
   _id : 0, 
   aReduced: {$reduce: {
       input: "$a", 
       initialValue: {x : 0, y: 1 }, 
       in: {x : {$sum: ["$$this.x", "$$value.x"] }, y: {$multiply: ["$$this.y", "$$value.y"] }   }    } } } 
   }
])
[ 
  { aReduced: { x: 8, y: 44 } }, 
  { aReduced: { x: 0, y: 1 } } 
]
```

Reply:
This is a fair example. It is not a matter of expressiveness or functionality, since a user wishing to obtain the same result is able to write the following reduction (3 lines below), which produces the above results:

```
// setup; array contains the documents
ArrayExpression<DocumentExpression> array = ofArray(
          of(Document.parse("{ a: [ { x: 3, y: 4 }, { x: 5, y: 11 } ] }")),
          of(Document.parse("{ a: [ ] }")));
DocumentExpression empty = of(Document.parse("{}"));

// reduction
Function<ArrayExpression<DocumentExpression>, DocumentExpression> multiReduce = (a) -> empty
          .setField("x", a.sum(e -> e.getInteger("x")))
          .setField("y", a.multiply(e -> e.getInteger("y")));

// assertion and get/set
assertExpression(
          Arrays.asList(
               Document.parse("{ aReduced: { x: 8, y: 44 } }"),
               Document.parse("{ aReduced: { x: 0, y: 1 } }")),
          array.map(d -> empty.setField("aReduced", multiReduce.apply(d.getArray("a")))));
```
It is, however, a performance issue - we want to perform those two reductions at the same time. For the sake of performance, I agree that we should offer a variant that allows multiple reductions to occur in tandem, which is what [$group](https://www.mongodb.com/docs/manual/reference/operator/aggregation/group/) does. That is, once we figure out the API for group, the problematic example you propose will be addressed, without the need to expose `reduce`.

1. We can always add reduce later but cannot remove it
2. Still possible to perform all operations using a less-error-prone syntax, though not as performantly
3. I believe we need to think more carefully about how to model the more unusual usages of `$reduce` in light of `$group` (note that `$reduce` is 37th by operation count)

Reply:
OK, I'm going to resolve this thread.  Please make sure that this is captured somewhere so it can be used to create documentation and generate subsequent Jira tickets.